### PR TITLE
Update decode.c

### PR DESCRIPTION
**Pull request by plolcott**
(1) Add a 32-bit stack in place of the current 16-bit stack so that the code is fully 32-bit.  
(2) Allow room for the NULL terminating byte. (One of my compilers caught this). 
I am running under Windows and Ubuntu.  I also adapted the code to compile as C++.

### CODE REVIEW

**Comment by plolcott** [comment]
I never could get test to build.

**Comment by plolcott** [comment]
Apparently it failed the 0008_seg_load_pm32.init regression test. 
I don't understand the libx86emu virtual memory system enough to proceed from here.

File: decode.c [hunk 1]
```diff
@@ -91,6 +91,7 @@ API_SYM unsigned x86emu_run(x86emu_t *emu, unsigned flags)
 
     if(ACC_D(emu->x86.R_CS_ACC)) {
       emu->x86.mode |= _MODE_DATA32 | _MODE_ADDR32 | _MODE_CODE32;
+	  emu->x86.mode |= _MODE_STACK32; // 2020-05-31 to be reviewed
```
Comment (wfeldt) [inline decode.c:94]: No, that's not correct. The CS selector's D bit provides only the default for code and ('normal') data accesses - in absense of a segment prefix. For stack accesses always the SS segment info is used. IIRC I did some experiments
on real hardware to get this implemented correctly back then.

I might be wrong, though. So, if you have a small code blob that works differently than the emulation, please post it here.

File: decode.c [hunk 2]
```diff
@@ -1844,7 +1845,7 @@ void log_regs(x86emu_t *emu)
 void check_data_access(x86emu_t *emu, sel_t *seg, u32 ofs, u32 size)
 {
   char **p = &emu->log.ptr;
-  static char seg_name[7] = "ecsdfg?";
+  static char seg_name[8] = "ecsdfg?";
```
Comment (wfeldt) [inline decode.c:1848]: The original length is correct. It's not a string, just an array of chars. And the code clearly limits the access to
index 6.

I see no warning in gcc 10, which is usually quite picky. Maybe your compiler has a way to tell it that here not a
asciiz-string is meant.

If I add the 0 byte just to shut a warning down, I might get a warning about an unused byte by some other
compiler later.

**Comment by plolcott** [comment]
The compiler always treats this code as a "c" string:
   static char seg_name[7] = "ecsdfg?";
and appends a  NULL terminating byte
I have shown proof of this below:

#include <cstdio>
#include <string>

static char seg_name[7] = "ecsdfg?";  // original code
static char seg_name1[]    = "ecsdfg?";
static char seg_name2[]    = {'e','c','s','d','f','g','?'};

// main() returns void to reduce clutter
void main(int argc, char *argv[])
{
//printf("sizeof(seg_name)=%d\n", sizeof(seg_name));
   printf("sizeof(seg_name1)=%d\n", sizeof(seg_name1));
   printf("sizeof(seg_name2)=%d\n", sizeof(seg_name2));
}

/***
compile
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27026.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.
test_seg_name.cpp
test_seg_name.cpp(4): error C2117: 'seg_name': array bounds overflow
test_seg_name.cpp(4): note: see declaration of 'seg_name'

run
D:\>test_seg_name
sizeof(seg_name1)=8
sizeof(seg_name2)=7
***/

------ Original Message ------

**Comment by plolcott** [comment]
I don't understand the virtual memory system very well at all especially
for stack usage.

When I ran the unmodified code the debug output seemed to indicate
that it was using a 16-bit stack. When I tested this code to force it to
overflow the 16-bit stack it failed indicating that it was a 16-bit 
stack.
It did a fixed number of recursive calls to the same function.

With the change that I made trying to force a 16-bit overflow seemed
to indicate that the stack was now in 32-bit mode.  I thought that it
would be a good idea to get this change reviewed because this was
the only change that I made that could have side-effects.

The change to the demo program adding 32-bit mode was enormously
helpful for my project. It would be really great if we could also get a
32-bit stack. I diligently strove to make absolute minimal changes to 
the
code base to transform the code so it would work as C++ under Windows
and Ubuntu.

------ Original Message ------

**Comment by wfeldt** [comment]
> I don't understand the virtual memory system very well at all especially for stack usage.

Actually, I misread your original [comment](https://github.com/wfeldt/libx86emu/issues/25#issuecomment-636510177). I assumed you were still talking about x86emu-demo.c.

So, I would be correct to also set the stack to 32bit in 32 bit mode here:

https://github.com/wfeldt/libx86emu/blob/master/demo/x86emu-demo.c#L161-L162

Your proposal in this pr is not correct. The INtel processor manual says about the `push` instruction:

> Decrements the stack pointer and then stores the source operand on the top of the stack. The address-size attribute of the stack segment determines the stack pointer size (16 bits or 32 bits), and the operand-size attribute of the current code segment determines the amount the stack pointer is decremented (2 bytes or 4 bytes). For example, if these address- and operand-size attributes are 32, the 32-bit ESP register (stack pointer) is decremented by 4 and, if they are 16, the 16-bit SP register is decremented by 2. (The B flag in the stack segment’s segment descriptor determines the stack’s address-size attribute, and the D flag in the current code segment’s segment descriptor, along with prefixes, determines the operand-size attribute and also the address-size attribute of the source operand.)

**Comment by plolcott** [comment]
Attached is the code that I used to test the stack depth.
It succeeds under my change and fails under your change.

------ Original Message ------

**Comment by wfeldt** [comment]
This is not the question. I agree that the demo code should also enable 32-bit mode for the stack segment if --32 is used.

What I'm saying is that your patch - which derives the stack layout type from the CS segment attributes - is wrong.

**Comment by plolcott** [comment]
Yes I understand. Mine is wrong because it breaks correct
emulation. We can tell because it flunks the regression test. Now we 
also have another test to see if the 32-bit stack is working.
The 32-bit stack test executes until completion of 50,000 instructions 
with my change that breaks emulation.

My next plan is to re-enable the regression tests on my code.  My code 
is currently based on static linking.

------ Original Message ------

**Comment by plolcott** [comment]
I didn't realize that you gave me the whole answer in your last reply. I 
added this at the spot that you highlighted:
emu->x86.R_SS_ACC |= (1 << 10);
I tested it with my test code and obviously it worked.

In the mean time I very carefully re-applied all of the changes that so 
that the library now compiles under Windows and Linux as "c" and "c++" 
and passes all the regression tests. I made sure to make the absolute 
minimum changes required for the port. This makes a DIFF show all the 
changes very cleanly.

The port to Windows mostly required making the Linux lib stuff invisible 
to Windows. It would only take a little more work to make this a Windows 
DLL under Windows. The port to C++ mostly required explicit casts.

Thanks again for all your help.

**Comment by plolcott** [comment]
I made the correct changes to the correct file this time.

I hardly know how github works at all so I may have updated the pull 
request incorrectly.

I did create another pull request but it didn't show up.

It looks like the system allowed me to directly merge to master. I 
didn't mean to do that.

**Comment by wfeldt** [comment]
> I made the correct changes to the correct file this time.

Kind of; there's still the original commit; and the file you changed got added to the wrong directory.

>I hardly know how github works at all so I may have updated the pull request incorrectly.

That's how it looks like.

> I did create another pull request but it didn't show up.

Yup, I didn't see any.

> It looks like the system allowed me to directly merge to master. I didn't mean to do that.

No, you did not merge into master here; possibly on your fork.

I can do a pr for the 32 bit one-line change, no problem. But if you have bigger changes in mind, like your C compiler adjustments, please do a clean new pull request.

**Comment by plolcott** [comment]
That is great, thanks. I know that making fewer changes is better because it reduces the chance that any change will be a breaking change. I went through the whole process again several times and there are only a few changes to get it to build as "C" in visual studio without errors or warnings. Getting rid of the errors only required hiding the Linux __attribute__ stuff from windows and using alternative port i/o functions. Getting rid of the warnings required adding some casts and turning off the warning related to strcat() safety. This will build as a visual studio project or from the command line batch file. I am proceeding to transforming this into building a Microsoft DLL.  As soon as I have this much done I will create another pull request.

**Pull request closed without merging**